### PR TITLE
Buffers

### PR DESCRIPTION
**Pull request by klemens-morgenstern**

### CODE REVIEW

File: include/boost/beast/core/buffer_ref.hpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
+//
+// Copyright (c) 2016-2019 Vinnie Falco (vinnie dot falco at gmail dot com)
```
Comment (vinniefalco) [inline include/boost/beast/core/buffer_ref.hpp:2]: if this is a new file put your name on it not mine..

File: include/boost/beast/core/static_buffer.hpp [hunk 2]
```diff
@@ -143,11 +143,10 @@ class static_buffer_base
     /** Returns a mutable buffer sequence representing writable bytes.
 
         Returns a mutable buffer sequence representing the writable
-        bytes containing exactly `n` bytes of storage. Memory may be
-        reallocated as needed.
+        bytes containing exactly `n` bytes of storage.
 
         All buffers sequences previously obtained using
-        @ref data or @ref prepare are invalidated.
+        @ref data or @ref prepare are maybe invalidated.
```
Comment (klemens-morgenstern) [inline include/boost/beast/core/static_buffer.hpp:149]:
```suggestion
        @ref data or @ref prepare are invalidated.
```

**Comment by vinniefalco** [comment]
is this going to be squashed?

File: include/boost/beast/core/buffer_ref.hpp [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+    /// Returns a mutable buffer sequence representing the readable bytes
+    mutable_buffers_type
+    data() noexcept
+    {
+        return buffer_.data();
+    }
+
+    /// @b DynamicBuffer_v1: Get a list of buffers that represents the output
+    /// sequence, with the given size.
```
Comment (vinniefalco) [inline include/boost/beast/core/buffer_ref.hpp:87]: Did you just copy this from asio?

**Comment by vinniefalco** [comment]
There is quite a lot of missing coverage, have you looked at the report?

File: include/boost/beast/core/buffer_ref.hpp (hunk 3 above)
Comment (klemens-morgenstern) [reply include/boost/beast/core/buffer_ref.hpp:87]: Yes.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,11 @@
+Version 340:
+
+* buffers have a features table in the docs
+* buffer_ref allows using beast buffers with asio
+* buffer_adaptor works with multi-buffer
```
Comment (vinniefalco) [inline CHANGELOG.md:5]: just squash the changelog commit into the last commit of the PR we don't need to see this clogging up the commit log

File: include/boost/beast/version.hpp [hunk 5]
```diff
@@ -20,7 +20,7 @@
    This is a simple integer that is incremented by one every
    time a set of code changes is merged to the develop branch.
 */
-#define BOOST_BEAST_VERSION 339
+#define BOOST_BEAST_VERSION 340
```
Comment (vinniefalco) [inline include/boost/beast/version.hpp:23]: This needs to be its own commit, "Set version to 340", and it should be at the top of the log

File: include/boost/beast/core/buffer_ref.hpp [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+
+template<class Allocator>
+class basic_flat_buffer;
+template<std::size_t N>
+class flat_static_buffer;
+template<class Allocator>
+class basic_multi_buffer;
+template<std::size_t N>
+class static_buffer;
```
Comment (vinniefalco) [inline include/boost/beast/core/buffer_ref.hpp:142]: Is doxygen going to pick these up?

Comment (klemens-morgenstern) [reply include/boost/beast/core/buffer_ref.hpp:142]: No.
